### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by Yoric**
This should already provide a first working version of the spam checker.

### CODE REVIEW

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    @staticmethod
+    def parse_config(config):
+        # No parsing needed
+        return config
+
+
+def _db_is_bad_link(db, table, link):
+    """
+    Search if any url in the database is a prefix of `link`.
+    `link` MUST be normalized by `_link_for_search`.
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:181]: `_link_for_search` doesn't seem to be defined.

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        # The table containing md5 hashes. Configured in homeserver.yaml, spam_checker.config.links_table.
+        self._md5_table = config["md5_table"]
+        logger.info("Using md5 table %s" % self._md5_table)
+
+        # The base url for this server. Configured in homeserver.yaml, spam_checker.config.base_url.
+        self._base_url = config["base_url"]
+        logger.info("Using base url %s" % self._base_url)
+
+        # Regexp for extracting info from mxc links.
+        self._mxc_re = re.compile("mxc://(?P<server_name>.*)/(?P<media_id>.*)")
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:41]: It seems that Synapse has a slightly different regex: https://github.com/matrix-org/synapse/blob/895e04319ba457a855207b8bb804f7360a258464/synapse/storage/databases/main/room.py#L663

Not sure if that's useful or not.

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    def user_may_create_room_alias(self, userid: str, room_alias: str) -> bool:
+        return True
+
+    def user_may_publish_room(self, userid: str, room_id: str) -> bool:
+        return True
+
+    @staticmethod
+    def parse_config(config):
+        # No parsing needed
+        return config
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:175]: I suspect this should do some assertions that the input are we expect (e.g. the parameters we want are there and they're a string).

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    row = db.fetchone()
+    if not row:
+        return False
+    return True
+
+
+# Run doctests
+if __name__ == "__main__":
+    import doctest
+    doctest.testmod()
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:212]: I don't think there are any tests, so this doesn't seem useful.

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    def user_may_publish_room(self, userid: str, room_id: str) -> bool:
+        return True
+
+    @staticmethod
+    def parse_config(config):
+        # No parsing needed
+        return config
+
+
+def _db_is_bad_link(db, table, link):
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:178]:
```suggestion
def _db_is_bad_link(db: str, table: str, link: str) -> bool:
```

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    row = db.fetchone()
+    if not row:
+        logger.info("No match in %s for link %s " % (table, link))
+        return False
+
+    # 2. Check whether it's actually a prefix.
+    logger.info("Located potential prefix %s" % row[0])
+    return link.startswith(row[0])
+
+def _db_is_bad_upload(db, table, md5):
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:198]:
```suggestion
def _db_is_bad_upload(db: str, table: str, md5: str) -> bool:
```

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+
+def _db_is_bad_link(db, table, link):
+    """
+    Search if any url in the database is a prefix of `link`.
+    `link` MUST be normalized by `_link_for_search`.
+    """
+    # Note: As per IWF guidelines, we're looking for *prefixes*. This might
+    # be slow. We're quickly going to have 1M+ links, so we need to find out
+    # whether this slows things down.
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:185]: I think you and @richvdh discussed this query so I'm going to believe that this works. 😄

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                media_id = match.group('media_id')
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:143]: That's a good question. 😢

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                except Exception as e:
+                    # In case of timeout or error, there's nothing we can do.
+                    # Let's not take the risk of blocking valid contents.
+                    logger.warn("Could not download media: %s" % e)
+                md5 = hashlib.md5()
+                await response.collect(lambda batch: md5.update(batch))
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:149]: I'm not really sure what this is doing -- `IResponse` / `Response` doesn't have a `collect` method?

Anyway, I'm hoping we can iteratively md5-hash things as you're suggesting here. 👍

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+                        return True
+
+        # If it's a file, download content, extract hash.
+        elif event["content"]["msgtype"] == "m.file":
+            if not await self.can_we_check_md5():
+                return False
+
+            match = self._mxc_re.match(event["content"]["url"])
+            server_name = None
+            media_id = None
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:131]: No reason to define these here.

File: synapse_spamcheck_badlist/BadListFilter.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
+                except Exception as e:
```
Comment (clokep) [inline synapse_spamcheck_badlist/BadListFilter.py:144]: This error checking does not seem to work, I think you want to `continue` after logging.

**Review by clokep — COMMENTED** [review]
Overall this looks pretty reasonable.

A few general notes:

* We usually put license headers at the top of each file. See https://github.com/matrix-org/synapse/blob/895e04319ba457a855207b8bb804f7360a258464/synapse/logging/context.py#L1-L14 I don't know who the proper copyright holder is to put in this file.
* Usually for Python projects you'll want a gitignore that includes `.pyc` and `*.egg` and `*.egg-info`.

File: synapse_spamcheck_badlist/BadListFilter.py (hunk 9 above)
Comment (Yoric) [reply synapse_spamcheck_badlist/BadListFilter.py:149]: But this is actually a `treq.response`, which does :)
https://treq.readthedocs.io/en/release-20.3.0/api.html#treq.response._Response.collect

File: synapse_spamcheck_badlist/BadListFilter.py (hunk 11 above)
Comment (Yoric) [reply synapse_spamcheck_badlist/BadListFilter.py:144]: I think we want to bailout and `return False`, no?

Comment (clokep) [reply synapse_spamcheck_badlist/BadListFilter.py]: I don't know what behavior is best, it depends if you want to fail closed or opened (and potentially if you want retries). You might also want to check other links.

File: synapse_spamcheck_badlist/BadListFilter.py (hunk 9 above)
Comment (clokep) [reply synapse_spamcheck_badlist/BadListFilter.py]: Interesting. I suspect we should update the type signature in Synapse then.

File: synapse_spamcheck_badlist/BadListFilter.py (hunk 11 above)
Comment (Yoric) [reply synapse_spamcheck_badlist/BadListFilter.py]: Other links? There can be more than one file in a `m.file` message?

Comment (clokep) [reply synapse_spamcheck_badlist/BadListFilter.py]: Oops, sorry I got this confused with checking the links in a standard message!

**Pull request merged**